# Thesis Update

# My first Optimized Classification Problem using DNN

### Importing Libraries

In [1]:
import tensorflow as tf
import numpy as np
import pandas as pd
#import sklearn

/home/ram/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/ram/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
/home/ram/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/home/ram/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


### Importing Keras Parameters

In [2]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam
from keras.utils import to_categorical

Using TensorFlow backend.


### Importing Data File (.csv)

In [3]:
df = pd.read_csv('sensor_readings_24.csv', delimiter = ',', dtype = np.str)

### Viewing Dataset

In [4]:
print(df)

      0.438  0.498  3.625  3.645  5.000  2.918 5.000.1  2.351  2.332  2.643  \
0     0.438  0.498  3.625  3.648  5.000  2.918   5.000  2.637  2.332  2.649   
1     0.438  0.498  3.625  3.629  5.000  2.918   5.000  2.637  2.334  2.643   
2     0.437  0.501  3.625  3.626  5.000  2.918   5.000  2.353  2.334  2.642   
3     0.438  0.498  3.626  3.629  5.000  2.918   5.000  2.640  2.334  2.639   
4     0.439  0.498  3.626  3.629  5.000  2.918   5.000  2.633  2.334  2.645   
5     0.440  5.000  3.627  3.628  5.000  2.919   3.028  2.346  2.330  2.638   
6     0.444  5.021  3.631  3.634  5.000  2.919   5.000  2.626  2.327  2.638   
7     0.451  5.025  3.635  3.639  5.000  2.920   3.027  2.620  2.323  2.632   
8     0.458  5.022  3.640  3.644  5.000  2.922   5.000  2.346  2.321  2.628   
9     0.465  0.525  3.646  3.670  5.000  2.923   5.000  2.611  2.315  2.631   
10    0.473  0.533  3.652  3.676  5.000  2.925   5.000  2.607  2.310  2.623   
11    0.481  0.544  3.658  3.678  5.000  2.926   5.0

### Cleaning Dataset (Converting String to Integer)

In [5]:
df = df.replace('Move-Forward',0)
df = df.replace('Slight-Right-Turn',1)
df = df.replace('Sharp-Right-Turn',2)
df = df.replace('Slight-Left-Turn',3)

### Converting Dataset into Matrix

In [6]:
data = pd.DataFrame.as_matrix(df)

### Verifying Matrix Dimensions

In [7]:
print(data.shape)

(5455, 25)


### Splitting Dataset into Training and Testing without using in-built Function

In [8]:
s = int(0.7*data.shape[0])

X_train = data[0:s,0:data.shape[1]-1]
X_test = data[s:data.shape[0],0:data.shape[1]-1]

y_train = data[0:s,data.shape[1]-1].astype(int)
y_test = data[s:data.shape[0],data.shape[1]-1].astype(int)

### Converting Classes in y to Categorical Variable

In [9]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

### Designing our model

In [10]:
model = Sequential()
model.add(Dense(28, input_dim = 24, activation = 'relu', kernel_initializer = 'he_uniform'))
model.add(Dense(24, activation = 'relu'))
model.add(Dense(20, activation = 'relu'))
model.add(Dense(16, activation = 'relu'))
model.add(Dense(12, activation = 'relu'))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(4, activation = 'softmax'))

### Compiling Model

In [11]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

### Fitting Model

In [13]:
model.fit(X_train, y_train, epochs=10, batch_size=10)

Epoch 1/10
3818/3818 [==============================] - 3s 896us/step - loss: 0.3477 - acc: 0.8669
Epoch 2/10
3818/3818 [==============================] - 3s 861us/step - loss: 0.3288 - acc: 0.8800
Epoch 3/10
3818/3818 [==============================] - 3s 847us/step - loss: 0.3141 - acc: 0.8845
Epoch 4/10
3818/3818 [==============================] - 3s 797us/step - loss: 0.3080 - acc: 0.8874
Epoch 5/10
3818/3818 [==============================] - 3s 868us/step - loss: 0.2976 - acc: 0.8882
Epoch 6/10
3818/3818 [==============================] - 3s 832us/step - loss: 0.2785 - acc: 0.9054
Epoch 7/10
3818/3818 [==============================] - 3s 873us/step - loss: 0.2814 - acc: 0.8979
Epoch 8/10
3818/3818 [==============================] - 3s 875us/step - loss: 0.2700 - acc: 0.8989
Epoch 9/10
3818/3818 [==============================] - 3s 872us/step - loss: 0.2635 - acc: 0.9049
Epoch 10/10
3818/3818 [==============================] - 3s 855us/step - loss: 0.2575 - acc: 0.9083


####  The dataset consists of distance measured by 24 ultrasonic sensors.
#### I optimized my DNN from 33% to 90% accuracy in just 10 epochs.
# -----------------------------------------------------------------------------------------------------------------------
##### Short-Term Memory Mechanisms in Neural NetworkLearning of Robot Navigation Tasks: A Case Study (https://www.researchgate.net/publication/224117546_Short-term_memory_mechanisms_in_neural_network_learning_of_robot_navigation_tasks_A_case_study)